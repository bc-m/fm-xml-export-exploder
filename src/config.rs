#[derive(Debug, Clone, Copy, PartialEq, Eq, Hash)]
pub enum CatalogType {
    Accounts,
    BaseDirectory,
    BaseTable,
    CalcsForCustomFunctions,
    CustomFunctions,
    CustomMenu,
    CustomMenuSet,
    ExternalDataSource,
    ExtendedPrivileges,
    FieldsForTables,
    FileAccess,
    Layout,
    Library,
    OptionsForValueLists,
    PrivilegeSets,
    Relationship,
    Script,
    StepsForScripts,
    TableOccurrence,
    Theme,
    ValueList,
}

impl CatalogType {
    pub fn from_bytes(bytes: &[u8]) -> Option<Self> {
        match bytes {
            b"AccountsCatalog" => Some(Self::Accounts),
            b"BaseDirectoryCatalog" => Some(Self::BaseDirectory),
            b"BaseTableCatalog" => Some(Self::BaseTable),
            b"CalcsForCustomFunctions" => Some(Self::CalcsForCustomFunctions),
            b"CustomFunctionsCatalog" => Some(Self::CustomFunctions),
            b"CustomMenuCatalog" => Some(Self::CustomMenu),
            b"CustomMenuSetCatalog" => Some(Self::CustomMenuSet),
            b"ExternalDataSourceCatalog" => Some(Self::ExternalDataSource),
            b"ExtendedPrivilegesCatalog" => Some(Self::ExtendedPrivileges),
            b"FieldsForTables" => Some(Self::FieldsForTables),
            b"FileAccessCatalog" => Some(Self::FileAccess),
            b"LayoutCatalog" => Some(Self::Layout),
            b"LibraryCatalog" => Some(Self::Library),
            b"OptionsForValueLists" => Some(Self::OptionsForValueLists),
            b"PrivilegeSetsCatalog" => Some(Self::PrivilegeSets),
            b"RelationshipCatalog" => Some(Self::Relationship),
            b"ScriptCatalog" => Some(Self::Script),
            b"StepsForScripts" => Some(Self::StepsForScripts),
            b"TableOccurrenceCatalog" => Some(Self::TableOccurrence),
            b"ThemeCatalog" => Some(Self::Theme),
            b"ValueListCatalog" => Some(Self::ValueList),
            _ => None,
        }
    }

    pub fn get_config(&self) -> CatalogConfig {
        match self {
            Self::Accounts => CatalogConfig {
                catalog_item_name: b"Account".to_vec(),
                out_folder_name: "accounts".to_string(),
                wrapped_in_object_list: true,
                uses_folders: false,
                id_path: "Account/AccountReference/@id".to_string(),
            },
            Self::BaseDirectory => CatalogConfig {
                catalog_item_name: b"BaseDirectory".to_vec(),
                out_folder_name: "base_directory".to_string(),
                wrapped_in_object_list: false,
                uses_folders: false,
                id_path: "".to_string(),
            },
            Self::BaseTable => CatalogConfig {
                catalog_item_name: b"BaseTable".to_vec(),
                out_folder_name: "table_stubs".to_string(),
                wrapped_in_object_list: false,
                uses_folders: false,
                id_path: "".to_string(),
            },
            Self::CalcsForCustomFunctions => CatalogConfig {
                catalog_item_name: b"CustomFunctionCalc".to_vec(),
                out_folder_name: "custom_function_calcs".to_string(),
                wrapped_in_object_list: true,
                uses_folders: false,
                id_path: "CustomFunctionCalc/CustomFunctionReference/@id".to_string(),
            },
            Self::CustomFunctions => CatalogConfig {
                catalog_item_name: b"CustomFunction".to_vec(),
                out_folder_name: "custom_function_stubs".to_string(),
                wrapped_in_object_list: true,
                uses_folders: true,
                id_path: "".to_string(),
            },
            Self::CustomMenu => CatalogConfig {
                catalog_item_name: b"CustomMenu".to_vec(),
                out_folder_name: "custom_menus".to_string(),
                wrapped_in_object_list: false,
                uses_folders: false,
                id_path: "".to_string(),
            },
            Self::CustomMenuSet => CatalogConfig {
                catalog_item_name: b"CustomMenuSet".to_vec(),
                out_folder_name: "custom_menu_sets".to_string(),
                wrapped_in_object_list: true,
                uses_folders: false,
                id_path: "".to_string(),
            },
            Self::ExternalDataSource => CatalogConfig {
                catalog_item_name: b"ExternalDataSource".to_vec(),
                out_folder_name: "external_data_sources".to_string(),
                wrapped_in_object_list: false,
                uses_folders: false,
                id_path: "".to_string(),
            },
            Self::ExtendedPrivileges => CatalogConfig {
                catalog_item_name: b"ExtendedPrivilege".to_vec(),
                out_folder_name: "extended_privileges".to_string(),
                wrapped_in_object_list: true,
                uses_folders: false,
                id_path: "".to_string(),
            },
            Self::FieldsForTables => CatalogConfig {
                catalog_item_name: b"FieldCatalog".to_vec(),
                out_folder_name: "table_fields".to_string(),
                wrapped_in_object_list: false,
                uses_folders: false,
                id_path: "FieldCatalog/BaseTableReference/@id".to_string(),
            },
            Self::FileAccess => CatalogConfig {
                catalog_item_name: b"Authorization".to_vec(),
                out_folder_name: "file_access".to_string(),
                wrapped_in_object_list: true,
                uses_folders: false,
                id_path: "".to_string(),
            },
            Self::Layout => CatalogConfig {
                catalog_item_name: b"Layout".to_vec(),
                out_folder_name: "layouts".to_string(),
                wrapped_in_object_list: false,
                uses_folders: true,
                id_path: "Layout/LayoutReference/@id".to_string(),
            },
            Self::Library => CatalogConfig {
                catalog_item_name: b"BinaryData".to_vec(),
                out_folder_name: "libraries".to_string(),
                wrapped_in_object_list: false,
                uses_folders: false,
                id_path: "BinaryData/LibraryReference/@id".to_string(),
            },
            Self::OptionsForValueLists => CatalogConfig {
                catalog_item_name: b"ValueList".to_vec(),
                out_folder_name: "value_list_options".to_string(),
                wrapped_in_object_list: false,
                uses_folders: false,
                id_path: "ValueList/ValueListReference/@id".to_string(),
            },
            Self::PrivilegeSets => CatalogConfig {
                catalog_item_name: b"PrivilegeSet".to_vec(),
                out_folder_name: "privilege_sets".to_string(),
                wrapped_in_object_list: true,
                uses_folders: false,
                id_path: "".to_string(),
            },
            Self::Relationship => CatalogConfig {
                catalog_item_name: b"Relationship".to_vec(),
                out_folder_name: "relationships".to_string(),
                wrapped_in_object_list: false,
                uses_folders: false,
                id_path: "".to_string(),
            },
            Self::Script => CatalogConfig {
                catalog_item_name: b"Script".to_vec(),
                out_folder_name: "script_stubs".to_string(),
                wrapped_in_object_list: false,
                uses_folders: true,
                id_path: "".to_string(),
            },
            Self::StepsForScripts => CatalogConfig {
                catalog_item_name: b"Script".to_vec(),
                out_folder_name: "script_steps".to_string(),
                wrapped_in_object_list: false,
                uses_folders: false,
                id_path: "Script/ScriptReference/@id".to_string(),
            },
            Self::TableOccurrence => CatalogConfig {
                catalog_item_name: b"TableOccurrence".to_vec(),
                out_folder_name: "table_occurrences".to_string(),
                wrapped_in_object_list: false,
                uses_folders: false,
                id_path: "".to_string(),
            },
            Self::Theme => CatalogConfig {
                catalog_item_name: b"Theme".to_vec(),
                out_folder_name: "themes".to_string(),
                wrapped_in_object_list: false,
                uses_folders: false,
                id_path: "".to_string(),
            },
            Self::ValueList => CatalogConfig {
                catalog_item_name: b"ValueList".to_vec(),
                out_folder_name: "value_list_stubs".to_string(),
                wrapped_in_object_list: false,
                uses_folders: false,
                id_path: "".to_string(),
            },
        }
    }
}

#[derive(Debug)]
pub struct CatalogConfig {
    pub catalog_item_name: Vec<u8>,
    pub out_folder_name: String,
    pub wrapped_in_object_list: bool,
    pub uses_folders: bool,
    pub id_path: String,
}

pub struct Flags {
    pub parse_all_lines: bool,
    pub lossless: bool,
    pub verbose: Option<Option<String>>,
}
