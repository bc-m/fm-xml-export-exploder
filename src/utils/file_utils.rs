use anyhow::{anyhow, Error, Result};
use std::path::PathBuf;

/// Validate that a path exists and is a directory
pub fn valid_dir_or_throw(dir_path: &PathBuf) -> Result<(), Error> {
    let metadata = std::fs::metadata(dir_path)
        .map_err(|_| anyhow!("Path '{}' not exists", dir_path.display()))?;

    match metadata.is_dir() {
        true => Ok(()),
        false => Err(anyhow!("Path '{}' is not a directory", dir_path.display())),
    }
}

/// Join scope ID and name into a formatted string
pub fn join_scope_id_and_name(scope_id: &str, scope_name: &str) -> String {
    format!("{} - ID {}", scope_name, scope_id)
}

/// Escape filename by replacing special characters with underscores
pub fn escape_filename(filename: &str) -> String {
    let escaped_filename = filename
        .chars()
        .map(|c| match c {
            '/' | '\\' | ':' | '*' | '?' | '"' | '<' | '>' | '|' | '{' => '_',
            _ => c,
        })
        .collect::<String>();

    escaped_filename.trim().to_string()
}

/// Check if a line should be skipped based on content
pub fn should_skip_line(line: &str) -> bool {
    if line.contains("<TagList></TagList>") {
        return true;
    };
    if line.contains("<OwnerID></OwnerID>") {
        return true;
    };
    if line.contains("<Options>0</Options>") {
        return true;
    }; // Default
    if line.contains("<Options>1048576</Options>") {
        return true;
    }; // Breakpoint
    if line.contains("<UUID>") && line.contains("</UUID>") {
        return true;
    };

    false
}
