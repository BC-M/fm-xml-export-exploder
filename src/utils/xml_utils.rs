use crate::utils::attributes::get_attributes;
use quick_xml::events::{BytesCData, BytesEnd, BytesStart, BytesText, Event};
use quick_xml::Reader;
use std::io::{BufRead, Read};

pub fn start_element_to_string(e: &BytesStart) -> String {
    let mut complete_tag = format!("<{}", local_name_to_string(e.name().as_ref()));
    for attr in get_attributes(e).unwrap() {
        // de-noise TODO: make optional
        match attr.0.as_str() {
            "nextvalue" | "UUID" | "index" => {
                continue;
            }
            _ => {}
        }

        complete_tag.push_str(&format!(" {}=\"{}\"", attr.0, attr.1));
    }
    complete_tag.push('>');

    complete_tag
}

pub fn cdata_element_to_string(e: &BytesCData) -> String {
    let mut content = String::new();
    content.push_str("<![CDATA[");
    content.push_str(cdata_to_string(e).as_str());
    content.push_str("]]>");

    content
}

pub fn text_element_to_string(e: &BytesText, escape: bool) -> String {
    if escape {
        decode_xml_special_characters(text_to_string(e))
    } else {
        text_to_string(e)
    }
}

pub fn end_element_to_string(e: &BytesEnd) -> String {
    let element_name = local_name_to_string(e.name().as_ref());
    format!("</{}>", element_name)
}

pub fn local_name_to_string(local_name: &[u8]) -> String {
    return match std::str::from_utf8(local_name) {
        Ok(text) => text.to_string(),
        Err(_) => String::new(),
    };
}

pub fn text_to_string(e: &BytesText) -> String {
    return match e.unescape() {
        Ok(text) => text.to_string(),
        Err(_) => String::new(),
    };
}

pub fn cdata_to_string(e: &BytesCData) -> String {
    return match std::str::from_utf8(e) {
        Ok(text) => text.to_string(),
        Err(_) => String::new(),
    };
}

fn decode_xml_special_characters(input: String) -> String {
    input
        .replace('&', "&amp;")
        .replace('\"', "&quot;")
        .replace('<', "&lt;")
        .replace('>', "&gt;")
        .replace('\'', "&apos;")
        .replace('\r', "&#13;")
}

pub fn skip_element<R: Read + BufRead>(reader: &mut Reader<R>, _: &BytesStart) {
    let mut depth = 1;
    let mut buf: Vec<u8> = Vec::new();
    loop {
        match reader.read_event_into(&mut buf) {
            Err(_) => continue,
            Ok(Event::Eof) => break,
            Ok(Event::Start(_)) => {
                depth += 1;
            }
            Ok(Event::End(_)) => {
                depth -= 1;
                if depth == 0 {
                    break;
                }
            }
            _ => {}
        }
        buf.clear();
    }
}

#[cfg(test)]
mod tests {
    use super::*;

    #[test]
    fn test_decode_xml_special_characters() {
        assert_eq!(
            decode_xml_special_characters("This & that \"test\" <tag>".to_string()),
            "This &amp; that &quot;test&quot; &lt;tag&gt;"
        );
    }
}
