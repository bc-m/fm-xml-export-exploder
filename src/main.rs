use std::{fs, path::PathBuf, time::Instant};

use anyhow::Result;
use clap::Parser;
use rayon::prelude::*;

use crate::config::Flags;
use crate::utils::file_utils::valid_dir_or_throw;
use crate::utils::xml_utils::XmlEventType;
use crate::xml_processor::explode_xml;

mod catalog;
mod config;
mod script_sanitizer;
mod script_steps;
mod supporting;
mod tests;
mod utils;
mod xml_processor;

/// Parse all as XML exported FileMaker solutions from source directory and explode them to target directory.
#[derive(Parser)]
#[command(version, about, long_about = None)]
struct Args {
    /// The source directory to read input
    source: PathBuf,

    /// The target directory to write output
    target: PathBuf,

    /// Parse all lines (or skip less important ones to reduce noise)
    #[arg(short, long)]
    all_lines: bool,

    /// Retain all information from the main xml (or skip less important catalogs and attributes)
    #[arg(short, long)]
    lossless: bool,

    /// Enable debug/verbose output. Optionally filter by catalog name.
    /// Example: --verbose "CustomFunctionsCatalog"
    #[arg(short, long, value_name = "FILTER", num_args(0..=1))]
    verbose: Option<Option<String>>,
}

#[derive(Debug, Default)]
pub struct Skeleton {
    pub content: String,
    pub previous_line: String,
    pub previous_event_type: XmlEventType,
}

fn main() -> Result<()> {
    let start = Instant::now();

    let args = Args::parse();
    let in_dir = args.source;
    let out_dir = args.target;
    let flags = Flags {
        parse_all_lines: args.all_lines,
        lossless: args.lossless,
        verbose: args.verbose,
    };

    match &flags.verbose {
        Some(Some(filter)) => println!("🔍 Verbose mode (filtered on {})", filter),
        Some(None) => println!("🔊 Verbose mode (all)"),
        None => (), // silent
    };

    valid_dir_or_throw(&in_dir)?;

    // Read directory contents
    let paths = fs::read_dir(in_dir)?
        .filter_map(|entry| entry.ok().map(|e| e.path())) // Filter out directories and unwrap results
        .filter(|path| path.is_file() && path.extension().unwrap_or_default() == "xml") // Filter XML files
        .collect::<Vec<_>>(); // Collect paths into a vector

    println!("Start processing {} files...", paths.len());

    // Process XML files in parallel
    paths.par_iter().for_each(|path| {
        match explode_xml(path, &out_dir, &flags) {
            Ok(_) => {}
            Err(err) => {
                let file_name = path.file_name().unwrap().to_str().unwrap();
                eprintln!("Failed to process file '{}': {}", file_name, err)
            }
        };
    });

    let duration = start.elapsed();
    if duration.as_secs() > 9 {
        println!("Completed in {:?} seconds.", duration.as_secs());
    } else {
        println!("Completed in {:?} ms.", duration.as_millis());
    }

    Ok(())
}
