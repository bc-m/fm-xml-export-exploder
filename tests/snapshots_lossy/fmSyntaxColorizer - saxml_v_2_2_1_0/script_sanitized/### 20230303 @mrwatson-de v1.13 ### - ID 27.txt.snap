---
source: src/main.rs
expression: output_content
---
# File: fmSyntaxColorizer

# © 2012-2023 MrW russell@mrwatson.de

# 20230303 @mrwatson-de v1.13
  - Added `and color`, `or color`, `xor color` and `not color`
# 20230303 @mrwatson-de v1.12
  - Added Markers + tipps to GUI to help guide the user
  - Added `comment color`
# 20230225 @mrwatson-de v1.11
  - Fixed #10 Duplicate Palette
  - Fixed #11 `field color` added - to same color as layout color
  - Fixed #12 Users can now read the Changes / Version Script
# 20230125 @mrwatson-de v1.10
  - Added Missing MBS Extras (Export Tooltips, Autocomplete Fn+Params, …)
# 20221228 @mrwatson-de v1.9.1
- Corrected FM-Version field names
# 20221208 @mrwatson-de v1.9
- Added Re-Login button (for quick administration)
# 20220927 MrW v1.8.4
  - Removed fmScriptWorkspace
# 20220922 MrW v1.8.3
  - Removed all GBS syntax rules which tried and failed to find commented out calculations
# 20220812 MrW v1.8.2
  - Fixed a false Syntax warning caused by an over-lax syntax rule which tried to find commented out calculations
# 20220318 MrW v1.8.1
  - [improvement] Changed the color highlighting to ScriptResult rather than ScriptParam
# 20220317 MrW v1.8 Improvements to Syntax-Colorizing:
  - [feature] Added *ScriptParam* syntax to help distinguish ScriptParameters from ScriptResults! (I've just used ScriptParameter once too often where I should be using ScriptResult!)
  - [feature] Pimped Markdown Headings (#...######) to make the structure of the script more vivid
# 20211209 MrW v1.7 Added MBS Variable autocomplete
# 20211208 MrW v1.6.1 Moved '#!' and '#/' to the Core syntax, whilst '#+Info+' and '#-Info-' reamin in GBS syntax
# 20211208 MrW v1.6 Added syntax and colors to mark comments as Important or unimportant
  - added colors 'Comment Important' and 'Comment Unimportant' and
  - added comment prefix '!' or '+Info+' to mark important comments
  - added comment prefix '/' or '-Info-' to mark unimportant comments
# 20211208 MrW v1.5.2 Made searching for categories easier
# 20211205 MrW v1.5.1 Improved "typical script" to show parameter definition in comments better
# 20211104 MrW v1.5 Added options to set line limits on MBS Variable Check script scanning
# 20211015 MrW v1.4.1 Corrected MBS Version number required for new formula wildcards
# 20211010 MrW v1.4 Added new colorizing functionality:
  - [feature] Added support for wildcard matches in calculation formulas, for example *zzz*()
  - [feature] Added zzz as warning to Core collection - I *think* this is probably general enough for the masses
  - [feature] Added MBS Blue color for MBS Script Step
  - [enhancement] Reduced the number of syntax definitions by deleting unnecessary entries
  - [enhancement] Also moved functions out of the Core to a new category, since they are all undefined in the default color palette anyhow
# 20211008 MrW v1.3.1 Fixed [bug] where colors were visible in the middle of the Color-Palettes layout
# 20210924 MrW v1.3 Improved support for Dark Mode (Issue [#6])
  - [feature] Added a new dark mode Palette from Kristian Olsen: KO-Colors (Dark)
  - [feature] Added radio button to identify Dark/Light-Mode Palettes
  - [feature] Improved dark mode display of colors and selected palette
  - [feature] Added a tip after switching or colorizing, explaining what to do when switching between light & dark modes
  - [feature] Automatically switch between suggested light/(Dark) palettes
# 20210916 MrW v1.2 You can simply change colors again with right-click > Text Color !
  - [bug] fixed the Format menu, so that users can again set colors using right click > Text Color - AT LAST! ([#1][i1])
  - [bug] removed some lingering pesky German  ([#3][i3])
  - [bug] fixed the release defaults to *not* turn on MBS calculation font size resizing to 30 px! ([#4][i4])
  
  [i1]: https://github.com/mrwatson-de/fmSyntaxColorizer/issues/1
  [i3]: https://github.com/mrwatson-de/fmSyntaxColorizer/issues/3
  [i4]: https://github.com/mrwatson-de/fmSyntaxColorizer/issues/4
# 20210803 MrW 1.1.2 Corrected latest MBS Extras
# 20210803 MrW 1.1.1 Added latest MBS Extras inc. 'Debugger: Tooltips', 'Help Menu Entries', 'Schema - Don't Open in Relation Graph', 'Calculations: Calc. Tools', 'Calculations: Match Brackets'
# 20210322 MrW 1.0.1 Fixed VersionNr function to work for FM19.2
# 20210114 MrW 1.0 fmSyntaxColorizer goes open source :-) Let's call it Version 1.0 at last!
# 20201106 MrW 0.9v52 Highlighted the Colorize button - better UX
# 20201104 MrW 0.9v51 Corrected User's script access privs
# 20200708 MrW 0.9v50 Added recognition of targeted variables (Target: $)
# 20200706 MrW 0.9v49 Added Collections and GitHub links
# 20200528 MrW 0.9v48 Added support for DarkMode, a Default DarkMode theme (needs tweaking) and some further color names in the color palettte. Plus fixed the jump from the Syntax list to the current color palette
# 20191123 MrW 0.9v47 Added support for latest MBS options including keep column widths and custom calulation font size, search box width, and the search, highlight, if and loop colors.
# 20191008 MrW 0.9v46 Added support for global variable color and smaller bigger color (<>)
# 20191003 MrW 0.9v45 Corrected editability of newer scripts
# 20191003 MrW 0.9v44 Added sorted steps for translation EN <-> IT
# 20190524 MrW 0.9v43 Syntax for FileMaker 18
# 20190524 MrW 0.9v42 Added FM-Version when duplicating a syntax
# 20190204 MrW 0.9v41 Enabled fmWorkMate to be called from AppleScript (privilege + added public API scripts to Activate fmWorkMate and open some important modules)
# 20190203 MrW 0.9v40 Integrated fmScriptWorkspace 
# 20190103 MrW 0.9v39 Added Syntax + Colors for some (FileMakery) Markdown in comments and some attempts to catch commented out code
# 20180509 MrW 0.9v38 Added fm17 syntax;
  plus a great new "FileMaker Enhanced" color palette as an extension of FileMaker's standard syntax colors
  plus removed some superceded stuff
# 20180506 MrW 0.9v37 Added fm17 script stuff
# 20171006 MrW 0.9v36 Added some functions + steps from the last few fm releases that were missing
# 20171004 MrW 0.9v35 Added more important FileMakerDoc @tags
# 20171004 MrW 0.9v34 Added Warning color for Go to Record when Exit after last is Off
# 20170927 MrW 0.9v33 Added many of the coming FileMakerDoc @tags
# 20170524 MrW 0.9v32 Made starting fmSyntaxColorizer simpler, more powerful and more beautiful than ever.
  All MBS developer extras can now be turned on & off from the start page.
  Visual graphic shows exactly which options are activated.
  Foregrounded fmSyntaxColorizers minimized mode, and extended it 4 (fantastic) dynamic syntax colors.
  Updated all Syntax for fm16.
  Added new color group "21 Plugin Steps"
# 20170519 MrW 0.9v31 Changed New Window, Select Window and Close Window to Layout colours for fm16.
# 20170518 MrW 0.9v30 GUI overhaul ready for fm16 + MBS Plugins multiple new developer extras
# 20160924 MrW 0.9v29 Preparing for FM15 compatability
    - Added FM-Version Filter which automatically selects only the entries which are relevant for the current version
    - Small view face-lift for the Syntax list view
# 20150912 MrW 0.9v28 Upgraded to fm14 platform
# 20140602 MrW 0.9v27 Help
# 20140509 MrW 0.9v26 Clearer state buttons for SyntaxColorizing, LineNumbers and VariableChecking - now also available in midimised view, and a few other GUI improvements.
# 20140430 MrW 0.9v25 New App-Menu, shortcut keys and fab new midimised state for active switching between palettes - cool!
# 20140429 MrW 0.9v24 Click the Sort heading with [SHIFT]+[CTRL]+[ALT] to perform a hard sort
# 20140429 MrW 0.9v23 Corrected bug, whereby more than 1-click start may have been necessary
# 20140425 MrW 0.9v22 MBS-Plugin check extended to recognise a disabled plugin (fmp12 files)
# 20140423 MrW 0.9v21 Major improvements in the GUI:
  - Major overhaul of the home page on the principle of less is more.
  - Editing of colors including drag & drop corrected, with standardised help info on the cyclone headings
  - The (i) Web info page now uses the full height of the monitor
# 20140423 MrW 0.9v20 fmSyntaxColorizer now correctly removes syntax colorizing
# 20140423 MrW 0.9v19 Improved SyntaxColorizing:
  - Added syntax to mark "*CurrentPrivilegeSetName*" as "Undesirable"
  - Changed color for "Halt Script" and "Exit Application" to "User Interaction"
  - Corrected color name for "#@"
  - Improved typical script to highlight more features
# 20140422 MrW 0.9v18 Support for MBS v4.2.0.04: Variable checking improved to ignore cut-off variable names [e.g.: $EndOfLin…]
# 20131217 MrW 0.9v17 Close file instead of hiding - and open web
# 20131216 MrW 0.9v16 New commands for fm13 (ScriptSteps) and new line number color function added
# 20131204 MrW 0.9v15 Variable-Checking added
# 20131119 MrW 0.9v14 Added new warning colour for seemingly commented out content
# 20131023 MrW 0.9v13 New Example Script "All script steps sorted by ID"
# 20131021 MrW 0.9v12 With separate enable/disable & status for Line Numbers
# 20131017 MrW 0.9v11 Support for MBS v3.5 and script line numbers :-) MBS-Plugin check greatly improved
# 20130926 MrW 0.9v10 Bug removed: if no import file available returns to layout
# 20130924 MrW 0.9v9 Hugely improved palette import & export, where the user can choose which palettes to import
# 20130923 MrW 0.9v8 Better introduction ; plus About *... Menu corrected and with Continue… menu item
# 20121202 MrW 0.9v7 Added fmp12 Commands; added App-Version to home page ; improved Color-Palettes layout ; Script Menu added
# 20121201 MrW 0.9v6 Programme and data improved for Windows, improved Script examples
# 20121126 MrW 0.9v5 Added functioning application Help, tidied layouts
# 20121109 MrW 0.9v4 it just goes on and on!
  - Enabled Status and toggle button on the Home Screen
  - Now with script-wildcards, variables and function highlighting
# 20121102 MrW 0.9v3 Wow!
  - Table view for palette-colours
  - Color-names reordered to make a more logical rainbow
  - Palette Sharing with Export/Import
  - Better MBS-integration
  - Simplified start page with one-click start!
  - <Error> tags introduced
  - Value list meta-values ++New++ and --None-- (cleverly) removed from the palettes color portal!
  - Comment (i) button added to Commands
# 20121031 MrW 0.9v2 Further developments:
  
  - Script and Calc types separated
  - Improved Help Menu with Send Feedback
  
  - Foreign commands corrected
  - Languages Table + WL
  
# 20121030 MrW 0.9v1 Lots of changes:
  - More finely tuned palettes, incl. Errors, Warnings, User Interactions, Ifs and Loops 
  - More finely tuned commands, incl. Errors, Warnings and User Interactions
  - Calculation colors, incl. seperate portal for calculation colors in the palette view
  - Palette-switching, (colors are now referenced by name)
  - Test-scripts vastly improved
  - GUI tidied up a bit.
# 20121025 MrW 0.9v0 Nice :-)


Aktuelles Script verlassen [ Hole( DateiName ) & ":	" & Trimme( Austauschen( Hole( ScriptName ) ; "#" ; "" ) ) ]
